File: packages/dotnet/src/Codec.Net/MapLoader.cs
// SPDX-License-Identifier: MIT
using System.Net.Http;
using System.Security.Cryptography;
using System.Text;

namespace Codec;

/// <summary>Options for <see cref="MapLoader.LoadAsync"/>.</summary>
public sealed class LoadOptions
{
    /// <summary>URL to fetch the map from.</summary>
    public required string Url { get; init; }

    /// <summary>
    /// Optional sha256 hex digest to verify the fetched map against.
    /// Accepts <c>sha256:&lt;hex&gt;</c> or bare <c>&lt;hex&gt;</c>. If omitted, no verification.
    /// </summary>
    public string? Hash { get; init; }

    /// <summary>Pluggable cache. Defaults to a static in-memory cache.</summary>
    public IMapCache? Cache { get; init; }

    /// <summary>HttpClient. If omitted, a static default is used.</summary>
    public HttpClient? Http { get; init; }

    /// <summary>Cache key. Defaults to <c>{Url}#{Hash}</c>.</summary>
    public string? CacheKey { get; init; }
}

/// <summary>Thrown when a fetched map doesn't match the expected hash.</summary>
public sealed class TokenizerMapHashMismatchException : Exception
{
    public string Expected { get; }
    public string Actual { get; }

    public TokenizerMapHashMismatchException(string expected, string actual)
        : base($"TokenizerMap hash mismatch.\n  expected: {expected}\n  actual:   {actual}")
    {
        Expected = expected;
        Actual = actual;
    }
}

/// <summary>Fetch, verify, and cache tokenizer maps.</summary>
public static class MapLoader
{
    private static readonly IMapCache DefaultCache = new MemoryMapCache();
    private static readonly HttpClient DefaultHttp = CreateDefaultHttp();

    private static HttpClient CreateDefaultHttp()
    {
        // Enable transparent decompression so brotli/gzip/zstd Content-Encoding
        // from CDNs (jsDelivr serves brotli for free) works without callers
        // having to configure anything.
        var handler = new HttpClientHandler
        {
            AutomaticDecompression =
                System.Net.DecompressionMethods.GZip
                | System.Net.DecompressionMethods.Brotli
                | System.Net.DecompressionMethods.Deflate,
        };
        var client = new HttpClient(handler);
        client.DefaultRequestHeaders.UserAgent.ParseAdd("Codec.Net/0.1");
        return client;
    }

    /// <summary>
    /// Fetch, verify, and cache a tokenizer map. Cache hits skip the network.
    /// </summary>
    public static async Task<TokenizerMap> LoadAsync(LoadOptions opts, CancellationToken ct = default)
    {
        var cache = opts.Cache ?? DefaultCache;
        var http = opts.Http ?? DefaultHttp;
        var cacheKey = opts.CacheKey ?? $"{opts.Url}#{opts.Hash ?? string.Empty}";

        var cached = await cache.GetAsync(cacheKey, ct).ConfigureAwait(false);
        if (cached is not null) return cached;

        using var resp = await http.GetAsync(opts.Url, HttpCompletionOption.ResponseHeadersRead, ct).ConfigureAwait(false);
        resp.EnsureSuccessStatusCode();
        var bytes = await resp.Content.ReadAsByteArrayAsync(ct).ConfigureAwait(false);

        if (opts.Hash is not null)
        {
            var expected = ParseHash(opts.Hash);
            var actual = Sha256Hex(bytes);
            if (!string.Equals(expected, actual, StringComparison.OrdinalIgnoreCase))
                throw new TokenizerMapHashMismatchException(expected, actual);
        }

        var map = TokenizerMap.FromJson(bytes);
        await cache.SetAsync(cacheKey, map, ct).ConfigureAwait(false);
        return map;
    }

    private static string Sha256Hex(byte[] bytes)
    {
        Span<byte> hash = stackalloc byte[32];
        SHA256.HashData(bytes, hash);
        var sb = new StringBuilder(64);
        foreach (var b in hash) sb.Append(b.ToString("x2"));
        return sb.ToString();
    }

    private static string ParseHash(string hash)
    {
        var colon = hash.IndexOf(':');
        if (colon < 0) return hash.ToLowerInvariant();
        var algo = hash[..colon].ToLowerInvariant();
        if (algo != "sha256")
            throw new NotSupportedException($"Unsupported hash algorithm: {algo} (only sha256 supported)");
        return hash[(colon + 1)..].ToLowerInvariant();
    }
}

File: packages/java/src/main/java/ai/codec/MapLoader.java
// SPDX-License-Identifier: MIT
package ai.codec;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.time.Duration;
import java.util.Locale;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;

/**
 * Fetch, verify, and cache tokenizer maps.
 *
 * <p>Maps are content-addressed: the URL is untrusted and the sha256 hash
 * is the trust anchor. Cache hits skip the network entirely.
 */
public final class MapLoader {
    private MapLoader() {}

    private static final MapCache DEFAULT_CACHE = new InMemoryMapCache();
    private static final HttpClient DEFAULT_HTTP = HttpClient.newBuilder()
            .connectTimeout(Duration.ofSeconds(30))
            .followRedirects(HttpClient.Redirect.NORMAL)
            .build();

    /** Synchronously fetch, verify, and cache a tokenizer map. Cache hits skip the network. */
    public static TokenizerMap load(LoadOptions opts) {
        try {
            return loadAsync(opts).get();
        } catch (Exception e) {
            Throwable c = e.getCause() != null ? e.getCause() : e;
            if (c instanceof TokenizerMapHashMismatchException tm) throw tm;
            if (c instanceof TokenizerMapValidationException tv) throw tv;
            if (c instanceof RuntimeException re) throw re;
            throw new RuntimeException(c);
        }
    }

    /** Asynchronously fetch, verify, and cache a tokenizer map. */
    public static CompletableFuture<TokenizerMap> loadAsync(LoadOptions opts) {
        MapCache cache = opts.cache() != null ? opts.cache() : DEFAULT_CACHE;
        HttpClient http = opts.http() != null ? opts.http() : DEFAULT_HTTP;
        String cacheKey = opts.cacheKey() != null
                ? opts.cacheKey()
                : (opts.url() + "#" + (opts.hash() != null ? opts.hash() : ""));

        Optional<TokenizerMap> cached = cache.get(cacheKey);
        if (cached.isPresent()) {
            return CompletableFuture.completedFuture(cached.get());
        }

        HttpRequest req = HttpRequest.newBuilder()
                .uri(URI.create(opts.url()))
                .header("Accept", "application/json")
                // gzip is the only encoding the JDK HttpClient handles transparently
                // when set explicitly via response decoder. We let the server pick.
                .header("Accept-Encoding", "gzip")
                .GET()
                .build();

        return http.sendAsync(req, HttpResponse.BodyHandlers.ofByteArray())
                .thenApply(resp -> {
                    int code = resp.statusCode();
                    if (code < 200 || code >= 300)
                        throw new RuntimeException("MapLoader: HTTP " + code + " from " + opts.url());

                    byte[] bytes = decodeBody(resp);

                    if (opts.hash() != null) {
                        String expected = parseHash(opts.hash());
                        String actual = sha256Hex(bytes);
                        if (!expected.equalsIgnoreCase(actual))
                            throw new TokenizerMapHashMismatchException(expected, actual);
                    }

                    TokenizerMap map = TokenizerMap.fromJson(bytes);
                    cache.set(cacheKey, map);
                    return map;
                });
    }

    /** Decode response body honoring optional gzip Content-Encoding. */
    private static byte[] decodeBody(HttpResponse<byte[]> resp) {
        Optional<String> enc = resp.headers().firstValue("content-encoding");
        byte[] body = resp.body();
        if (enc.isEmpty()) return body;
        String e = enc.get().toLowerCase(Locale.ROOT);
        if (e.equals("gzip")) {
            try (java.util.zip.GZIPInputStream gz = new java.util.zip.GZIPInputStream(new java.io.ByteArrayInputStream(body));
                 java.io.ByteArrayOutputStream out = new java.io.ByteArrayOutputStream()) {
                byte[] tmp = new byte[8 * 1024];
                int n;
                while ((n = gz.read(tmp)) > 0) out.write(tmp, 0, n);
                return out.toByteArray();
            } catch (IOException ioe) {
                throw new RuntimeException("MapLoader: gzip decode failed: " + ioe.getMessage(), ioe);
            }
        }
        // Other encodings (br, zstd, deflate) are not handled by the JDK HttpClient;
        // we don't request them above.
        return body;
    }

    static String sha256Hex(byte[] bytes) {
        try {
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            byte[] hash = md.digest(bytes);
            StringBuilder sb = new StringBuilder(64);
            for (byte b : hash) sb.append(String.format("%02x", b));
            return sb.toString();
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException("SHA-256 unavailable", e);
        }
    }

    static String parseHash(String hash) {
        int colon = hash.indexOf(':');
        if (colon < 0) return hash.toLowerCase(Locale.ROOT);
        String algo = hash.substring(0, colon).toLowerCase(Locale.ROOT);
        if (!algo.equals("sha256"))
            throw new UnsupportedOperationException("Unsupported hash algorithm: " + algo + " (only sha256 supported)");
        return hash.substring(colon + 1).toLowerCase(Locale.ROOT);
    }
}
